Sources/SwiftGRPCNIO/HTTP1ToRawGRPCClientCodec.swift

```
@@ -0,0 +1,153 @@
+/// reads length-prefxied gRPC messages into `ByteBuffer`s containing serialized
+/// Protobuf messages.
+///
+/// The outbound processing transforms serialized Protobufs into length-prefixed
+/// gRPC messages stored in `ByteBuffer`s.
+///
+/// See `HTTP1ToRawGRPCServerCodec` for the corresponding server codec.
+public final class HTTP1ToRawGRPCClientCodec {
+  public init() {}
+
+  private enum State {
+    case expectingHeaders
+    case expectingBodyOrTrailers
+    case ignore
+  }
+
+  private var state: State = .expectingHeaders
+  private let messageReader = LengthPrefixedMessageReader(mode: .client)
+  private let messageWriter = LengthPrefixedMessageWriter()
+  private var inboundCompression: CompressionMechanism = .none
+}
+
+extension HTTP1ToRawGRPCClientCodec: ChannelInboundHandler {
+  public typealias InboundIn = HTTPClientResponsePart
+  public typealias InboundOut = RawGRPCClientResponsePart
+
+  public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    if case .ignore = state { return }
+
+    switch unwrapInboundIn(data) {
+    case .head(let head):
+      state = processHead(ctx: ctx, head: head)
+
+    case .body(var message):
+      do {
+        state = try processBody(ctx: ctx, messageBuffer: &message)
+      } catch {
+        ctx.fireErrorCaught(error)
+        state = .ignore
+      }
+
+    case .end(let trailers):
+      state = processTrailers(ctx: ctx, trailers: trailers)
+    }
+  }
+
+  /// Forwards the headers from the request head to the next handler.
+  ///
+  /// - note: Requires the `.expectingHeaders` state.
+  private func processHead(ctx: ChannelHandlerContext, head: HTTPResponseHead) -> State {
+    guard case .expectingHeaders = state
+      else { preconditionFailure("received headers while in state \(state)") }
+
+    if let encodingType = head.headers["grpc-encoding"].first {
+      inboundCompression = CompressionMechanism(rawValue: encodingType) ?? .unknown
```

Should we already fail here in case of an unexpected compression mechanism?